treewide: Use deprecations in code rather than comments

chrysn opened the pull request:
This is a replacement PR for #18565  because that had been made to the wrong branch and GitHub closed it irrevertably when I tried to clean up.

The relevant parts of that PR are replicated here:

---

### Contribution description

This executes <https://github.com/RIOT-OS/RIOT/issues/18561>.

It already has the macros and Doxygen stuff, it lacks
* change of all `@deprecated` to `DEPRECATED`
* CI time rejection of non-code deprecation (unsure if possible, maybe we still need it for module level deprecations)
* Removal of deprecated functions in the code base, or annotation with `-Wno-error=deprecated-declarations` if we still want to test the deprecated functions.

### Testing procedure

Right now:

* See how CI starts breaking because we -Werror and the one function that got changed is still used in tests.
* Look at Doxygen output on the changed function.
* [edit: added] Also try building an example of a harder deprecation with a newer C standard where the actual C version (and not the GCC proprietary one) is used:
```
$ CFLAGS=-std=c2x make -C examples/usbus_minimal BOARD=samr21-xpro
```

[edit: or the more forgiving gnu2x, given that native has some trouble with being precisely conformant]

Long term:

* Try using a deprecated function, see things fail.

---

@maribu said:

If you throw in an `-Wno-error=deprecated-declarations`, the warning should not be escalated to an error. We may however want to still escalate it to an error with `RIOT_CI_BUILD=1` to avoid in-tree users sneaking back in.

--

@chrysn responded:

On the severity of errors, I think a good path would be to have Makefiles of `ALLOW_DEPRECATED ?= 1`, make Murdock set `ALLOW_DEPRECATED = 0`, and have some tests (those that explicitly include deprecated functions because we don't want these to accidentally break) set `ALLOW_DEPRECATED = 1`.

---

On the CI side, Murdock went well because it duly rejected the tests that used the deprecated function (even though, as above, these should later be exempt).

---

Sadly, doxygen output [breaks completely](https://output.circle-artifacts.com/output/job/e2b7691b-ea06-4bbb-a7ed-f16e0c77353f/artifacts/0/doc/group__drivers__periph__flashpage.html) with the current doxygen version -- deprecation text gets moved into functions list, but the function itself is not present any more at all, and the next function is somehow also missing ... putting this in as "waiting for CI update" because the next Ubuntu hopefully has a better Doxygen. (Things looked good with the 1.9.4 I have here).

This is now blocked by https://github.com/RIOT-OS/riotdocker/pull/189

chrysn committed:
core: Add DEPRECATE(reason) macro

Contributes-To: https://github.com/RIOT-OS/RIOT/issues/18561

chrysn committed:
drivers/flashpage: Use in-code deprecation annotation

Contributes-To: https://github.com/RIOT-OS/RIOT/issues/18561

maribu commented:
I think this is @chrysn's plan anyway. But for the CI runs it will be an error (except for some apps explicitly testing for backward compatibility), so that no new in-tree users sneak in.

chrysn commented:
Yes, that's the plan.

---

Docs now look as good as it gets: [flashpage_first_free looks like that](https://output.circle-artifacts.com/output/job/7dd9a850-f337-4f44-a20c-3ca8808b75f9/artifacts/0/doc/group__drivers__periph__flashpage.html#ga8142b0ed43506621a76b573a61f5c90b) 
![Screenshot 2022-09-27 at 07-33-29 Flash page driver](https://user-images.githubusercontent.com/442326/192440583-0a3776c7-3a18-412c-84e7-7c5fcafca6f2.png)
(i.e. no links, and quotes around the text). If that's not good enough, I think we'd need to resort to double-tracking -- having text both in the docs and in the deprecation, and keeping them in sync.

Reviewers, is that good enough for you? (I'd like to know *before* I do the treewide edits).

chrysn committed:
makefiles: Add FORBID_DEPRECATED parameter, set in CI

This keeps deprecations at the warning level during regular development,
but enforces that no deprecated items are used in internal code, unless
the test explicitly allows that.

chrysn commented:
I've made the intended handling of deprecations explicit in the latest commit. It introduces a `FORBID_DEPRECATED` variable similar to `WERROR`. The code that maps the variable to the `CFLAGS` peeks into WERROR as well in order not to flood the CFLAGS with redundant input.

The new behavior is as follows, with the strongest rules on top:
* Particular tests, such as the one changed, can force `FORBID_DEPRECATED=0` to allow testing things until they are eventually removed.
* Murdock sets `FORBID_DEPRECATED=1` as a general preference.
* In absence of any settings, the use of deprecated items is allowed (and produces a compiler warning)

Do you have any preferences as to whether the treewide edit (moving as many deprecations as possible into C attributes) would be in here or a follow-up (with this one merely setting the stage)?

chrysn committed:
fixup! drivers/flashpage: Use in-code deprecation annotation

Add missing kernel_defines header

chrysn committed:
suit/coap: Use in-code deprecation annotation

chrysn committed:
usbus: Use in-code deprecation annotation

This change also serves as an example of how an item can still be used
in code that needs to use it for some time after the deprecation.

chrysn commented:
While I'm adding selected examples, I've started looking into how to get The Big List of things that should be deprecated in C rather than in documentation. Right now my best expression is

```
$ rg @deprecated -g '!*.txt' -g '!vendor'
```

because doc.txt files and vendor headers often include Doxygen deprecations that can't (practically) be moved to C.

I also found that there's no straightforward way to deprecated enums as a whole -- a `DEPRECATED("...")` line before an `enum {` does create the right documentation, but to make the individual values raise, there'd need to be a deprecation marker after every single value (eg. `SPI_NODEV DEPRECATED("...") = -ENXIO,`). That's impractical right now (both from a maintenance and a documentation side), so I'll leave those aside for now. (Also, that *concrete* deprecation would trigger quite a lot of errors during CI, given how often it still occurs in the code). Worse goes for groups of defines, which AFAICT can not be recognizably deprecated at all.

One more case of "code needs to declare when it uses deprecated stuff" I've found is that when a deprecated enum value (such as `USBDEV_EVENT_TR_FAIL`) is C-level deprecated, implementing a case for it (which may easily be necessary during the deprecation period) also trips. This can be remedied by pushing pragmas (which contrary to the literal value works with all our supported compilers); an example is now pushed. If we don't like that, we'd probably just relax on how deprecations become hard in CI (either per example/test, or generally).

In the end, replacing the existing documentation deprecations with C ones is likely to be a highly manual process, with many case-by-case decisions; so I'm leaning toward moving that to a later commit (except for the few items done so far, which'd serve as examples).

chrysn committed:
fixup! usbus: Use in-code deprecation annotation

Thanks @bergzand for clarifying.

chrysn committed:
fixup! usbus: Use in-code deprecation annotation

maribu commented on sys/include/suit/transport/coap.h:
```diff
@@ -27,17 +27,16 @@
- * @deprecated This is an alias for @ref suit_worker_run and will be removed
- *             after after the 2023.01 release.
  */
+DEPRECATED("This is an alias for suit_worker_run and will be removed after after the 2023.01 release.")
```
```suggestion
DEPRECATED("This is an alias for suit_worker_run and will be removed after "
           "the 2023.01 release.")
```

maribu commented on sys/include/suit/transport/coap.h:
```diff
@@ -89,12 +88,10 @@ extern const coap_resource_subtree_t coap_resource_subtree_suit;
  * @param[in] url       url pointer containing the full coap url to the manifest
  * @param[in] len       length of the url
  */
+DEPRECATED("This is an alias for suit_worker_trigger and will be removed after after the 2023.01 release.")
```
```suggestion
DEPRECATED("This is an alias for suit_worker_trigger and will be removed "
           "after the 2023.01 release.")
```

maribu reviewed (commented):
Looks good to me. If possible, I would prefer to split the lines to match the maximum char limit (e.g. see suggestions), so that the CI is happy. But I do not insist.

chrysn commented on core/lib/include/kernel_defines.h:
```diff
@@ -280,6 +280,28 @@ extern "C" {
+ * `[[deprecated(reason)]]`; otherwise to the GNUism
+ * `__((attribute(deprecated))`, or is ignored.
+ *
+ * In Doxygen, it is rendered as a deprecation; note that currently, no Doxygen
+ * commands work in the reason.
```
It does translate into something that works, but if that x were to contain, say, `Use @ref foo instead`, that foo would not be linked like in regular Doxygen comments.

maribu replied:
Ah, now I understand the wording. Could you change it to "in the reason text provided as argument"? I failed to understand what `reason` referred to; maybe others have the same issue.

chrysn replied:
It's "the text" now, with an example.

chrysn commented on drivers/include/periph/usbdev.h:
```diff
@@ -209,13 +210,10 @@ typedef enum {
      * An endpoint should emit this event after a nack reply to the host.
      */
-    USBDEV_EVENT_TR_FAIL,
+    USBDEV_EVENT_TR_FAIL
+    DEPRECATED("This event is deprecated as only a limited number of devices implement and doesn't offer value for upper layer code. This event will be removed after the 2022.07 release."),
```
That'd be nice, but sadly Doxygen doesn't take these as string literals (which would be concatenated) but as markdown, and would thus show the inner quotation marks at the line breaks.

What does work is making it a multiline string
```suggestion
    USBDEV_EVENT_TR_FAIL
    DEPRECATED("This event is deprecated as only a limited number of devices \
    implement and doesn't offer value for upper layer code. This event will be \
    removed after the 2022.07 release."),
```

but is that even generally legal C? (But maybe it doesn't need to be, as it only needs to be valid preprocessor / attribute...)

chrysn replied:
It works both with -std=c11 and -std=c2x, so let's go that way :-)

chrysn commented:
I'll need to rename this due to collisions -- `cpu/stellaris_common/include/vendor/` do some `ifndef DEPRECATED`, and that could conflict subtly.

Anyway, I'm not fully convinced that the deprecation with attributes works well around enum items, so my next change series will change this to:

* `RIOT_DEPRECATED_FUNCTION("text")` to go before function declarations,
* `RIOT_DEPRECATED_VARIANT(ENU_ITEM, "text")` to go around enum items

and apply line breaking.

chrysn committed:
squash!

Rename DEPRECATED to RIOT_DEPRECATED_x and split between enums and
functions to gain flexibility in where the attribute goes precisely (as
the documentation goes somewhere different)

This also word-wraps all the texts.

MrKevinWeiss commented:
Really nice, this is a good solution for C code, I wonder if we can expand this or at least reuse the marco semantics to handle deprecation of tools or build system info?

We would also need to publicise it to all maintainers.

MrKevinWeiss reviewed (commented):
Should we also handle compiler abstractions?  Maybe check for LLVM?

I suppose everything is gcc first and only portions of llvm are supported so maybe that would be OK since it is more about warnings rather then code execution.

maribu commented:
> Should we also handle compiler abstractions? Maybe check for LLVM?

Note that clang also defined `__GNUC__` and is *mostly* a drop-in replacement for GCC. The deprecated attribute is also supported by it: https://clang.llvm.org/docs/AttributeReference.html#deprecated

Since we don't support any compiler but GCC and clang/LLVM and there are no plans to do so, IMO this is just fine as it is.

maribu commented on core/lib/include/kernel_defines.h:
```diff
@@ -280,6 +280,61 @@ extern "C" {
 /* end{code-style-ignore} */
 
+/** @brief Declare a function as deprecated
+ *
```
```suggestion
 *
 * @param reason    Reason for removal
 * @param eol       Expected end of life
 * @param migration Migration path (e.g. API to use instead)
 *
```

Maybe enforce consistency and completeness of the doc by having explicit arguments for the important details. That would also decrease the itch to use fancy formatting.

Beware: Indent levels are not correct, did this on a phone.

maribu commented:
@chrysn Ping?
